framework/src/meshgenerators/ConcentricCircleMeshGenerator.C

```diff
@@ -326,51 +370,299 @@ ConcentricCircleMeshGenerator::generate()
 
-  counter = 0;
   while (index < limit)
   {
-
     Elem * elem = mesh->add_elem(new Quad4);
     elem->set_node(0) = nodes[index];
-    elem->set_node(1) = nodes[index + _num_sectors + 1];
-    elem->set_node(2) = nodes[index + _num_sectors + 2];
+    elem->set_node(1) = nodes[index + standard + 1];
+    elem->set_node(2) = nodes[index + standard + 2];
     elem->set_node(3) = nodes[index + 1];
 
-    for (int i = 0; i < static_cast<int>(subdomainIDs.size()) - 1; ++i)
+    for (int i = 0; i < static_cast<int>(subdomainIDs.size() - 1); ++i)
       if (index < limit - (standard + 1) * i && index >= limit - (standard + 1) * (i + 1))
         elem->subdomain_id() = subdomainIDs[subdomainIDs.size() - 1 - i];
 
-    int const initial = Utility::pow<2>(standard / 2 + 1);
-    int const final = Utility::pow<2>(standard / 2 + 1) + _num_sectors - 1;
+    const int initial = Utility::pow<2>(standard / 2 + 1);
+    const int final = Utility::pow<2>(standard / 2 + 1) + standard - 1;
 
     if ((index - initial) % (standard + 1) == 0)
       boundary_info.add_side(elem, 0, 2);
     if ((index - final) % (standard + 1) == 0)
       boundary_info.add_side(elem, 2, 1);
-    if (index > limit - (standard + 1))
+    if (!_has_outer_square)
+      if (index >= limit - (standard + 1))
+        boundary_info.add_side(elem, 1, 3);
+
+    ++index;
+    if (index == (num_nodes_boundary + counter * (standard + 1)) - 1)
+    {
+      ++index;
+      ++counter;
+    }
+  }
+
+  // adding elements for the moderator. (top left)
+  int initial =
+      Utility::pow<2>(standard / 2 + 1) + (standard + 1) * total_concentric_circles.size();
+
+  int initial2 =
+      Utility::pow<2>(standard / 2 + 1) + (standard + 1) * total_concentric_circles.size();
+
+  if (_has_outer_square)
+  {
+    if (_rings.back() != 0) // this must be condition up front.
     {
-      if (_has_outer_square)
+      index = Utility::pow<2>(standard / 2 + 1) + (standard + 1) * total_concentric_circles.size();
+      limit = Utility::pow<2>(standard / 2 + 1) + (standard + 1) * total_concentric_circles.size() +
+              (_rings.back() + 1) * (standard / 2) + Utility::pow<2>(_rings.back() + 2) - 3 -
+              _rings.back() * (_rings.back() + 2) - (_rings.back() + 1);
+      while (index <= limit)
       {
-        if (index < limit - standard + standard / 2)
+        Elem * elem = mesh->add_elem(new Quad4);
+        elem->set_node(0) = nodes[index];
+        elem->set_node(1) = nodes[index + 1];
+        elem->set_node(2) = nodes[index + 1 + _rings.back() + 1];
+        elem->set_node(3) = nodes[index + 1 + _rings.back()];
+        elem->subdomain_id() = subdomainIDs.back() + 1;
+
+        if (index < (initial2 + static_cast<int>(_rings.back())))
+          boundary_info.add_side(elem, 0, 1);
+
+        if (index == initial)
+          boundary_info.add_side(elem, 3, 4);
+
+        ++index;
+        if ((index - initial) % static_cast<int>(_rings.back()) == 0)
```

Comment:
I see a comment above that says that several of these types are `int`s due to modulo. `unsigned int` works fine as long as you don't attempt to feed negative values into the expression. It's a little unclear if your expressions are always positive so this might be fine. Just pointing it out in case you don't like all of the casts that are necessary.

Reply:
I agree. However, although both 'index', 'initial' are positive, '(index - intial)' can be negative, which causes an issue performing the modulo functions. If there are better methods to present this, please let me know. I will make changes correspondingly.